File: samples/HelloWorld/quicklua/QAtlas.lua
--[[/*
 * (C) 2012-2013 Marmalade.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */--]]

--------------------------------------------------------------------------------
-- QAtlas
--------------------------------------------------------------------------------
QAtlas = {}
QAtlas.__index = QAtlas

-- Override QNode GC function (still call old one at the end)
QAtlas.oldGC = getmetatable(quick.QAtlas).__gc
QAtlas.newGC = function(n)
    if n.name then
        dbg.print("GC on atlas: " .. n.name)
    else
        dbg.print("GC on atlas: (unnamed)")
    end
    QAtlas.oldGC(n)
end

--------------------------------------------------------------------------------
-- Private API
--------------------------------------------------------------------------------
QAtlas.serialize = function(o)
	local obj = serializeTLMT(getmetatable(o), o)
	return obj
end

--[[
/*
Initialise the peer table for the C++ class QAtlas.
This must be called immediately after the QAtlas() constructor.
*/
--]]
function QAtlas:initAtlas(n)
    local lp = {}
    setmetatable(lp, QAtlas)
    tolua.setpeer(n, lp)

    local mt = getmetatable(n) 
    mt.__serialize = QAtlas.serialize
    if config.debug.traceGC == true then
        mt.__gc = QAtlas.newGC
    end
end

--------------------------------------------------------------------------------
-- Public API
--------------------------------------------------------------------------------
--[[
/**
Create a texture atlas, specifying arbitrary input values.
@param values Lua table specifying name/value pairs, the path to a .plist file or the path to a texture.
@return The created atlas.
*/
--]]
function director:createAtlas(values)
    dbg.assertFuncVarTypes({"table", "string"}, values)
    local n = quick.QAtlas()
    QAtlas:initAtlas(n)

    if (type(values) == "table") then

        dbg.assert( values.textureName ~= nil, "director:createAtlas called without a valid textureName")
        n:initTexture( values.textureName)

        -- Calculate the scaling values for dynamic resolution support
        local texture_width, texture_height = n:getTextureSize()
        local source_width = values.originalSheetWidth or texture_width
        local source_height = values.originalSheetWidth or texture_width
        local scale_x = texture_width / source_width
        local scale_y = texture_width / source_width

        if values.frames ~= nil then
            -- initialise from a frames array
            for i,frame in ipairs(values.frames) do
                dbg.assert( frame.x ~= nil and frame.y ~= nil and frame.width ~= nil and frame.height ~= nil, "Each frame must contain an x, y, width and height member")

                local scaled_width = frame.width * scale_x
                local scaled_height = frame.height * scale_y

                n:addSpriteFrame( frame.x * scale_x, frame.y * scale_y,
                                    scaled_width, scaled_height,
                                    false,
                                    0, 0, scaled_width, scaled_height)

            end
        else
            -- initialise from a width/height/numFrames process
            dbg.assert( values.width ~= nil and values.height ~= nil and values.numFrames ~= nil, "Atlas simple parameter initialisation needs at least width, height and numFrames")

            -- Step the sprites at native resolution
            local sx = values.xStart or 0
            local x = sx
            local y = values.yStart or 0
            local remaining_frames = values.numFrames

            local border = values.border or 0
            local coverage_width = values.width + (border * 2)
            local coverage_height = values.height + (border * 2)

            local scaled_width = values.width * scale_x
            local scaled_height = values.height * scale_y

            -- Loop across and down the texture adding frames
            repeat
                repeat
                    n:addSpriteFrame( (x + border) * scale_x, (y + border) * scale_y,
                                       scaled_width, scaled_height,
                                       false,
                                       0, 0, scaled_width, scaled_height)

                    x = x + coverage_width

                    remaining_frames = remaining_frames - 1

                until (x + coverage_width) > texture_width or remaining_frames == 0

                x = sx
                y = y + coverage_height

            until (y + coverage_height) > texture_height or remaining_frames == 0

            dbg.assert( remaining_frames == 0, "The texture was too small for the number of images specified")

        end

    else
        -- Load via a path
        dbg.assertFuncVarTypes({"string"}, values)
        n:initFromFile(values)
    end

    -- Store this Atlas in the current QScene object's atlas list
--    dbg.print( "Adding atlas to scene "..tostring(self.currentScene))
    self.currentScene._atlasList[n] = n

    return n
end

File: source/QAtlas.cpp
/*
 * (C) 2012-2013 Marmalade.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

//#include "AppDelegate.h"
#include "QAtlas.h"
#include "QLuaHelpers.h"

#include "cocos2d.h"
#include <stdlib.h>

//For GNUC compiler warnings. Cocos2d-x macro which isnt defined in our current version of CC
#ifndef CC_UNUSED
#ifdef __GNUC__
#define CC_UNUSED __attribute__ ((unused))
#else
#define CC_UNUSED
#endif
#endif

USING_NS_CC;

QUICK_NAMESPACE_BEGIN;

//------------------------------------------------------------------------------
// QAtlas
//------------------------------------------------------------------------------
QAtlas::QAtlas()
{
	m_Texture = NULL;
    m_AutoGenerateNameIndex = 1;

    // Default blend mode
    _blendSrc = GL_ONE;
    _blendDst = GL_ONE_MINUS_SRC_ALPHA;
}
//------------------------------------------------------------------------------
QAtlas::~QAtlas()
{
    // Release our handle on the atlas texture
    if (m_Texture)
        CCTextureCache::sharedTextureCache()->removeTexture(m_Texture);
	CC_SAFE_RELEASE(m_Texture);

//    QTrace("QAtlas destroyed");
}
//------------------------------------------------------------------------------
bool QAtlas::initFromFile(std::string fileName)
{
    size_t pos = fileName.rfind('.');

    if (pos == std::string::npos)
        return false;

    std::string ext(fileName, pos);
    bool result = true;
    if (stricmp( ext.c_str(), ".plist") == 0)
    {
        // Call  the base class initialisation
        init();
        addSpriteFramesWithFile( fileName.c_str());
    }
    else
        result = InitFromTexture( fileName);

    // We need to set m_Texture
    if (!m_Texture)
    {
//        CCSpriteFrame* pSF = (CCSpriteFrame*)m_pSpriteFrames->objectForKey(0);
        CCSpriteFrame* pSF = this->GetFrame(0);
        m_Texture = pSF->getTexture();
    }

    // cocos2d::CCTexture must have pointer back to QAtlas
    QAssert(m_Texture, "Failed to set CCTexture for QAtlas");
    m_Texture->m_uID = (unsigned long)(void*)this;
    return result;
}
//------------------------------------------------------------------------------
bool QAtlas::initTexture(std::string filename)
{
    // Call  the base class initialisation
    init();

	CC_SAFE_RELEASE(m_Texture);
	m_Texture = CCTextureCache::sharedTextureCache()->addImage(filename.c_str());
    QAssert(m_Texture, "Failed to set CCTexture for QAtlas");
    m_Texture->retain();

    // cocos2d::CCTexture must have pointer back to QAtlas
    m_Texture->m_uID = (unsigned long)(void*)this;
    return true;
}
//------------------------------------------------------------------------------
void QAtlas::addSpriteFrame( float x, float y, float w, float h, bool rotated,
                        float ox, float oy, float sw, float sh)
{
    QAssert(m_Texture != NULL, "You must call initTexture before calling addSpriteFrame");

    CCSpriteFrame* frame = new CCSpriteFrame();

    frame->initWithTexture( m_Texture, 
				            CCRectMake(x, y, w, h), 
							rotated,
                            CCPointMake(ox, oy),
                            CCSizeMake(sw, sh)
							);

    AddSpriteWithAutogenName(frame);
}
//------------------------------------------------------------------------------
void QAtlas::getTextureSize(float *width, float *height)
{
    if (m_Texture != NULL)
    {
        CCSize texture_size = m_Texture->getContentSizeInPixels();

        *width = texture_size.width;
        *height = texture_size.height;
    }
}
//------------------------------------------------------------------------------
CCSpriteFrame* QAtlas::GetFrame(int frame) const
{
    QAssert(frame >=0 && frame < (int)m_pSpriteFrames->count(), "Tried to access an out of bounds sprite frame");

    CCString *key = (CCString *)(m_pSpriteFrames->allKeys()->objectAtIndex(frame));

    return (CCSpriteFrame*)m_pSpriteFrames->objectForKey( key->getCString());
}
//------------------------------------------------------------------------------
CCSpriteFrame* QAtlas::GetFrame( std::string frame) const
{
    CCSpriteFrame* pframe = (CCSpriteFrame*)m_pSpriteFrames->objectForKey( frame.c_str());

    QAssert(pframe != NULL, "Unable to find named frame");

    return pframe;
}
//------------------------------------------------------------------------------
static char* FormatAssertMsg(char* dest, const char* format, ...)
{
    va_list ap;
    va_start(ap, format);
    vsnprintf(dest, kMaxLogLen, format, ap);
    va_end(ap);
    return dest;
}
//------------------------------------------------------------------------------
void QAtlas::setTextureParams(const char* minFilter, const char* magFilter, const char* wrapS, const char* wrapT)
{
    // Default parameters
    ccTexParams cctp = 
    {
        GL_LINEAR_MIPMAP_NEAREST,
        GL_LINEAR,
        GL_CLAMP_TO_EDGE,
        GL_CLAMP_TO_EDGE,
    };

    struct NameValue { const char* s; int e; };

    // MIN FILTER
    static const NameValue minFilters[] =
    {
        {"GL_NEAREST_MIPMAP_NEAREST", GL_NEAREST_MIPMAP_NEAREST},
        {"GL_LINEAR_MIPMAP_NEAREST", GL_LINEAR_MIPMAP_NEAREST},
        {"GL_NEAREST_MIPMAP_LINEAR", GL_NEAREST_MIPMAP_LINEAR},
        {"GL_LINEAR_MIPMAP_LINEAR", GL_LINEAR_MIPMAP_LINEAR},
        {"GL_NEAREST", GL_NEAREST},
        {"GL_LINEAR", GL_LINEAR},
        {0, -1},
    };
    const NameValue* pNV = minFilters;
    while(pNV->e != -1)
    {
        if (!strcmp(minFilter, pNV->s))
        {
            cctp.minFilter = pNV->e;
            break;
        }
        pNV++;
    }
    char CC_UNUSED msg[kMaxLogLen + 1];
    QAssert(pNV->s, FormatAssertMsg(msg, "Unrecognised minFilter type: %s", minFilter));

    // MAG FILTER
    static const NameValue magFilters[] =
    {
        {"GL_NEAREST", GL_NEAREST},
        {"GL_LINEAR", GL_LINEAR},
        {0, -1},
    };
    pNV = magFilters;
    while(pNV->e != -1)
    {
        if (!strcmp(magFilter, pNV->s))
        {
            cctp.magFilter = pNV->e;
            break;
        }
        pNV++;
    }
    QAssert(pNV->s, FormatAssertMsg(msg, "Unrecognised magFilter type: %s", magFilter));

    // WRAP S
    static const NameValue wraps[] =
    {
        {"GL_REPEAT", GL_REPEAT},
        {"GL_CLAMP_TO_EDGE", GL_CLAMP_TO_EDGE},
        {"GL_MIRRORED_REPEAT", GL_MIRRORED_REPEAT},
        {0, -1},
    };

    if (!wrapS)
        cctp.wrapS = GL_CLAMP_TO_EDGE;
    else
    {
        pNV = wraps;
        while(pNV->e != -1)
        {
            if (!strcmp(wrapS, pNV->s))
            {
                cctp.wrapS = pNV->e;
                break;
            }
            pNV++;
        }
        QAssert(pNV->s, FormatAssertMsg(msg, "Unrecognised wrap type: %s", wrapS));
    }

    if (!wrapT)
        cctp.wrapT = GL_CLAMP_TO_EDGE;
    else
    {
        pNV = wraps;
        while(pNV->e != -1)
        {
            if (!strcmp(wrapT, pNV->s))
            {
                cctp.wrapT = pNV->e;
                break;
            }
            pNV++;
        }
        QAssert(pNV->s, FormatAssertMsg(msg, "Unrecognised wrap type: %s", wrapT));
    }

    // Set parameters on Cocos2d-x texture 
    m_Texture->setTexParameters(&cctp);
}
//------------------------------------------------------------------------------
void QAtlas::setBlendFunc(const char* blendSrc, const char* blendDst)
{
    // Default
    ccBlendFunc ccbf =
    {
        GL_ONE,
        GL_ONE,
    };

    struct NameValue { const char* s; int e; };
    static const NameValue funcs[] =
    {
        {"GL_ZERO", GL_ZERO},
        {"GL_ONE", GL_ONE},
        {"GL_SRC_COLOR", GL_SRC_COLOR},
        {"GL_ONE_MINUS_SRC_COLOR", GL_ONE_MINUS_SRC_COLOR},
        {"GL_DST_COLOR", GL_DST_COLOR},
        {"GL_ONE_MINUS_DST_COLOR", GL_ONE_MINUS_DST_COLOR},
        {"GL_SRC_ALPHA", GL_SRC_ALPHA}, 
        {"GL_ONE_MINUS_SRC_ALPHA", GL_ONE_MINUS_SRC_ALPHA},
        {"GL_DST_ALPHA", GL_DST_ALPHA},
        {"GL_ONE_MINUS_DST_ALPHA", GL_ONE_MINUS_DST_ALPHA},
        {"GL_CONSTANT_COLOR", GL_CONSTANT_COLOR},
        {"GL_ONE_MINUS_CONSTANT_COLOR", GL_ONE_MINUS_CONSTANT_COLOR},
        {"GL_CONSTANT_ALPHA", GL_CONSTANT_ALPHA},
        {"GL_ONE_MINUS_CONSTANT_ALPHA", GL_ONE_MINUS_CONSTANT_ALPHA},
        {"GL_SRC_ALPHA_SATURATE", GL_SRC_ALPHA_SATURATE},
        {0, -1},
    };
    const NameValue* pNV = funcs;
    while(pNV->e != -1)
    {
        if (!strcmp(blendSrc, pNV->s))
        {
            ccbf.src = pNV->e;
            break;
        }
        pNV++;
    }
    char CC_UNUSED msg[kMaxLogLen + 1];
    QAssert(pNV->s, FormatAssertMsg(msg, "Unrecognised blend function: %s", blendSrc));

    pNV = funcs;
    while(pNV->e != -1)
    {
        if (!strcmp(blendDst, pNV->s))
        {
            ccbf.dst = pNV->e;
            break;
        }
        pNV++;
    }
    QAssert(pNV->s, FormatAssertMsg(msg, "Unrecognised blend function: %s", blendDst));

    // Cocos2d-x stores these values on CCSprite etc.
    // We will store them directly on the QAtlas object, and copy across to CCSprite etc.
    // TODO - USING THESE VALUES FOR SPRITES ETC.
    _blendSrc = ccbf.src;
    _blendDst = ccbf.dst;
}

//------------------------------------------------------------------------------
// Private API
//------------------------------------------------------------------------------
bool QAtlas::InitFromTexture(std::string fileName)
{
    if (!initTexture( fileName))
        return false;

    CCSpriteFrame *spriteFrame = new CCSpriteFrame();
    spriteFrame->initWithTexture( m_Texture, 
                                    CCRectMake( 0.0f, 0.0f, (float)m_Texture->getPixelsWide(), (float)m_Texture->getPixelsHigh()), 
						            false,
                                    CCPointMake( 0.0f, 0.0f),
                                    CCSizeMake( (float)m_Texture->getPixelsWide(), (float)m_Texture->getPixelsHigh())
						            );

    AddSpriteWithAutogenName(spriteFrame);

    return true;
}
//------------------------------------------------------------------------------
void QAtlas::AddSpriteWithAutogenName(cocos2d::CCSpriteFrame* frame)
{
    // Create a unique number string for this sprite
    char number_buffer[8];
    number_buffer[7] = 0;
    char *number_ptr = &number_buffer[7];

    int index = m_AutoGenerateNameIndex++;
    while (index > 0)
    {
        number_ptr--;
        *number_ptr = (index % 10) + '0';
        index /= 10;
    }

    // Create a unique name for this spriteframe
    char name_buffer[16];
    strcpy(name_buffer, "QUICK_");
    strcat(name_buffer, number_ptr);

    // Add the spriteframe to the cache
    CCSpriteFrameCache::addSpriteFrame(frame, name_buffer);

    // Now the spriteframe has been put in the cache, we can release our handle
    frame->release();
}

QUICK_NAMESPACE_END;
